OSDOCS-11444: Adds low latency to MicroShift (#80236)
openshift/openshift-docs: modules/microshift-low-latency-prepare-workload.adoc

// Module included in the following assemblies:
//
// * microshift_configuring/microshift-low-latency.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-low-latency-prepare-workload_{context}"]
= Preparing a {microshift-short} workload for low latency

To take advantage of low latency, workloads running on {microshift-short} must have the `microshift-low-latency` container runtime configuration set using the `RuntimeClass` feature. The CRI-O `RuntimeClass` object is installed with the `microshift-low-latency` RPM, so only the pod annotations need to be configured.

.Prerequisites
* You installed the `microshift-low-latency` RPM package.

.Procedure

> . Use the following example to set the following annotations in the pod spec:

Reviewer: Do we want to list all available pod annotations and what they do, or we want to link to that information somewhere else?

Reviewer 2: I guess we can base the info on the enhancement (https://github.com/openshift/enhancements/blob/master/enhancements/microshift/low-latency-workloads-on-microshift.md#preparing-low-latency-workload) and supplement with info from ocp:
- https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-configuring-high-priority-workload-pods_cnf-provisioning-low-latency
- https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk_cnf-provisioning-low-latency
- https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-disabling-cpu-cfs-quota_cnf-provisioning-low-latency
- https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-disabling-interrupt-processing-for-individual-pods_cnf-provisioning-low-latency

Author: * Added the two additional fields

* Extrapolated what we want from https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk_cnf-provisioning-low-latency and put into the doc itself. 

* Linked the other three locations for user reading if they want to do so.

+
[source,text]
----
cpu-load-balancing.crio.io: "disable"
irq-load-balancing.crio.io: "disable"
cpu-quota.crio.io: "disable"
----
+
.Example pod that runs `oslat` test:
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: oslat
  annotations:
    cpu-load-balancing.crio.io: "disable" <1>
    irq-load-balancing.crio.io: "disable" <2>
    cpu-quota.crio.io: "disable" <3>
spec:
  runtimeClassName: microshift-low-latency <4>
  containers:
  - name: oslat
    image: quay.io/container-perf-tools/oslat
    imagePullPolicy: Always
    resources:
      requests:

>         memory: "400Mi"
>         cpu: "10"
>       limits:
>         memory: "400Mi"
>         cpu: "10"

Reviewer 2: Because we use `2,4-5` in the blueprint example, let's change it to cpu: 2

```suggestion
        memory: "400Mi"
        cpu: "2"
      limits:
        memory: "400Mi"
        cpu: "2"
```

    env:
    - name: tool
      value: "oslat"
    - name: manual
      value: "n"
    - name: PRIO
      value: "1"
    - name: delay
      value: "0"
    - name: RUNTIME_SECONDS
      value: "60"
    - name: TRACE_THRESHOLD
      value: ""
    - name: EXTRA_ARGS
      value: ""
    securityContext:
      privileged: true
      capabilities:
        add:
          - SYS_NICE
          - IPC_LOCK

----
<1> Disables the CPU load balancing for the pod.
<2> Opts the pod out of interrupt handling (IRQ).
<3> Disables the CPU completely fair scheduler (CFS) quota at the pod run time.
<4> The `runtimeClassName` must match the name of the performance profile configured in the cluster.
+
[IMPORTANT]
====
For the pod to have the `Guaranteed` QoS class, it must have the same values of CPU and memory in requests and limits. See link:https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#guaranteed[Guaranteed] (Kubernetes upstream documentation)
====

> . To apply changes, restart {microshift-short}.

Reviewer: Which changes would we be applying by restarting MicroShift for the workload?

Reviewer 2: We should remove this. At this point all the configurations requiring reboot or service restart should be completed

// Module included in the following assemblies:
//
// * microshift_configuring/microshift-low-latency.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-low-latency-prepare-workload_{context}"]
= Preparing a {microshift-short} workload for low latency

To take advantage of low latency, workloads running on {microshift-short} must have the `microshift-low-latency` container runtime configuration set using the `RuntimeClass` feature. The CRI-O `RuntimeClass` object is installed with the `microshift-low-latency` RPM, so only the pod annotations need to be configured.

.Prerequisites
* You installed the `microshift-low-latency` RPM package.

> //Does microshift need to be stopped?

Reviewer 2: No, these are part of the Pod definition

    imagePullPolicy: Always
    resources:
      requests:
        memory: "400Mi"
        cpu: "10"
      limits:
        memory: "400Mi"
        cpu: "10"
    env:
    - name: tool
      value: "oslat"
    - name: manual
      value: "n"
    - name: PRIO
      value: "1"
    - name: delay
      value: "0"
    - name: RUNTIME_SECONDS
      value: "60"
    - name: TRACE_THRESHOLD
      value: ""
    - name: EXTRA_ARGS
      value: ""
    securityContext:
      privileged: true
      capabilities:
        add:
          - SYS_NICE
          - IPC_LOCK

----
<1> Disables the CPU load balancing for the pod.
<2> Opts the pod out of interrupt handling (IRQ).
<3> Disables the CPU completely fair scheduler (CFS) quota at the pod run time.
<4> The `runtimeClassName` must match the name of the performance profile configured in the cluster.
+
[IMPORTANT]
====
For the pod to have the `Guaranteed` QoS class, it must have the same values of CPU and memory in requests and limits. See link:https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#guaranteed[Guaranteed] (Kubernetes upstream documentation)
====

. To apply changes, restart the workload pods.

> //Do we need to restart microshift and cri-o, too?

Reviewer 2: No, we don't need to restart workloads as well. If we create the manifest, they will be created. Not sure if we could update the manifest to have these settings required to have annotations and Guaranteed QoS - some fields cannot be changed in already running Pod

// Module included in the following assemblies:
//
// * microshift_configuring/microshift-low-latency.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-low-latency-prepare-workload_{context}"]
= Preparing a {microshift-short} workload for low latency

> To take advantage of low latency, workloads running on {microshift-short} must have the `microshift-low-latency` container runtime configuration set using the `RuntimeClass` feature. The CRI-O `RuntimeClass` object is installed with the `microshift-low-latency` RPM, so only the pod annotations need to be configured.

Reviewer 3: ```suggestion
To take advantage of low latency, workloads running on {microshift-short} must have the `microshift-low-latency` container runtime configuration set by using the `RuntimeClass` feature. The CRI-O `RuntimeClass` object is installed with the `microshift-low-latency` RPM, so only the pod annotations need to be configured.
```

.Prerequisites
* You installed the `microshift-low-latency` RPM package.
* You configured workload partitioning.

.Procedure

. Use the following example to set the following annotations in the pod spec:
+
[source,text]
----
cpu-load-balancing.crio.io: "disable"
irq-load-balancing.crio.io: "disable"
cpu-quota.crio.io: "disable"
cpu-load-balancing.crio.io: "disable"
cpu-freq-governor.crio.io: "<governor>"
----
+
.Example pod that runs `oslat` test:
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: oslat
  annotations:
    cpu-load-balancing.crio.io: "disable" # <1>
    irq-load-balancing.crio.io: "disable" # <2>
    cpu-quota.crio.io: "disable" # <3>

>     cpu-load-balancing.crio.io: "disable" # <4>

Reviewer 2: Judging by comment for <4> I think this should be something different
```suggestion
    cpu-c-states.crio.io: "disable" # <4>
```

    cpu-freq-governor.crio.io: "<governor>" # <5>
spec:
  runtimeClassName: microshift-low-latency # <6>
  containers:
  - name: oslat
    image: quay.io/container-perf-tools/oslat
    imagePullPolicy: Always
    resources:
      requests:
        memory: "400Mi"
        cpu: "2"
      limits:
        memory: "400Mi"
        cpu: "2"
    env:
    - name: tool
      value: "oslat"
    - name: manual
      value: "n"
    - name: PRIO
      value: "1"
    - name: delay
      value: "0"
    - name: RUNTIME_SECONDS
      value: "60"
    - name: TRACE_THRESHOLD
      value: ""
    - name: EXTRA_ARGS
      value: ""
    securityContext:
      privileged: true
      capabilities:
        add:
          - SYS_NICE
          - IPC_LOCK

----
<1> Disables the CPU load balancing for the pod.
<2> Opts the pod out of interrupt handling (IRQ).
<3> Disables the CPU completely fair scheduler (CFS) quota at the pod run time.

> <4> Enable or disable C-states for each CPU. Set the value to "disable" to provide the best performance for a high-priority pod.

Reviewer 3: For consistency in the list.
```suggestion
<4> Enables or disables C-states for each CPU. Set the value to "disable" to provide the best performance for a high-priority pod.
```

> <5> Sets the `cpufreq` governor for each CPU. The "performance" governor is recommended for high-priority workloads.

Reviewer 3: ```suggestion
<5> Sets the `cpufreq` governor for each CPU. The `performance` governor is recommended for high-priority workloads.
```

> <4> Enable or disable C-states for each CPU. Set the value to "disable" to provide the best performance for a high-priority pod. Disabling or enabling CPU load balancing is implemented on the CRI-O level using this field and the `runtimeClassName` field value.

Reviewer 2: It's also true for other annotations, they all are implemented by CRIO and all need specific runtimeClassName. If we want to mention that CRI-O is doing the heavylifting here maybe we can do in the section's intro?
```suggestion
<4> Enable or disable C-states for each CPU. Set the value to "disable" to provide the best performance for a high-priority pod.
```

<6> The `runtimeClassName` must match the name of the performance profile configured in the cluster. For example, `microshift-low-latency`.
+
[NOTE]
====

> Only disable CPU load balancing when the CPU manager static policy is enabled and for pods with guaranteed QoS that use whole CPUs. Otherwise, disabling CPU load balancing can affect the performance of other containers in the cluster.

Reviewer 3: Place _only_ as close to the thing it modifies as possible. 
```suggestion
Disable CPU load balancing only when the CPU manager static policy is enabled and for pods with guaranteed QoS that use whole CPUs. Otherwise, disabling CPU load balancing can affect the performance of other containers in the cluster.
```

> <6> The `runtimeClassName` must match the name of the performance profile configured in the cluster. For example, `microshift-low-latency`. Using the `microshift-low-latency` for `runtimeClassName` and setting `cpu-load-balancing.crio.io` to "disable" disables the CPU load balancing for the pod.

Reviewer 2: ```suggestion
<6> The `runtimeClassName` must match the name of the performance profile configured in the cluster. For example, `microshift-low-latency`.
```

Author: This is per the OCP docs. Is it not true for MicroShift?

Reviewer 2: It's true - I just felt it's out of place and redundant - we have <1> which described disabling cpu load balancing so I don't think we need to mention it in <6>.

Maybe the difference is OCP doesn't describe all annotation all at once so they need to make connection between this particular annotation and its requirement?

Author: OCP describes in it the docs, but theres a lot of extraneous information in that section about the NTO that doesn't apply to MicroShift. So I wanted to just make the admin aware that they have to use both of these fields to get the job done. See https://docs.openshift.com/container-platform/4.16/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-node-tuning-operator-disabling-cpu-load-balancing-for-dpdk_cnf-provisioning-low-latency

Author: I didn't want to link there and create confusion.

Reviewer 2: Yes, but it makes sense in case of OCP: they're focusing on specific label in that section, whereas we list all of the possible labels and the requirement of using right runtimeClassName is common to all of them.
I just don't see how providing extra info for just one of them is helping - it might lead something thinking that they don't need runtimeClass if they're not using cpu-load-balancing label.
